File: sdk/cpp/examples/blockchain.cpp
// Blockchain operations with the HyperMesh C++ SDK.
//
// Build: g++ -std=c++17 -I../include blockchain.cpp ../src/*.cpp ../src/api/*.cpp -o blockchain
// Run:   ./blockchain

#include <hypermesh/client.hpp>
#include <iostream>

int main() {
    hypermesh::HyperMeshClient client;

    // Get chain height
    auto h = client.blockchain().height();
    std::cout << "Blockchain height: " << h.height << "\n";

    // Fetch the genesis block
    auto genesis = client.blockchain().block(0);
    std::cout << "\nGenesis block:\n";
    std::cout << "  Index: " << genesis.index << "\n";
    std::cout << "  Hash: " << genesis.hash << "\n";
    std::cout << "  Previous hash: " << genesis.previous_hash << "\n";
    std::cout << "  Timestamp: " << genesis.timestamp << "\n";

    // Fetch the latest block
    if (h.height > 0) {
        auto latest = client.blockchain().block(h.height - 1);
        std::cout << "\nLatest block (index " << latest.index << "):\n";
        std::cout << "  Hash: " << latest.hash << "\n";
    }

    // Validate the chain
    auto result = client.blockchain().validate();
    std::cout << "\nBlockchain valid: " << (result.valid ? "true" : "false") << "\n";
    std::cout << "Blocks checked: " << result.blocks_checked << "\n";
    if (!result.errors.empty()) {
        std::cout << "Errors:\n";
        for (const auto& err : result.errors) {
            std::cout << "  - " << err << "\n";
        }
    }

    return 0;
}

File: sdk/typescript/examples/blockchain.ts
/**
 * Blockchain operations with the HyperMesh TypeScript SDK.
 *
 * Run: npx tsx examples/blockchain.ts
 */
import { HyperMeshClient } from "../src/index.js";

async function main() {
  const client = new HyperMeshClient();

  // Get the chain height
  const { height } = await client.blockchain.height();
  console.log("Blockchain height:", height);

  // Fetch the genesis block
  const genesis = await client.blockchain.block(0);
  console.log("\nGenesis block:");
  console.log("  Index:", genesis.index);
  console.log("  Hash:", genesis.hash);
  console.log("  Previous hash:", genesis.previous_hash);
  console.log("  Timestamp:", genesis.timestamp);

  // Fetch the latest block (if height > 0)
  if (height > 0) {
    const latest = await client.blockchain.block(height - 1);
    console.log("\nLatest block (index", latest.index, "):");
    console.log("  Hash:", latest.hash);
  }

  // Validate the chain
  const validation = await client.blockchain.validate();
  console.log("\nBlockchain valid:", validation.valid);
  if (validation.blocks_checked !== undefined) {
    console.log("Blocks checked:", validation.blocks_checked);
  }
  if (validation.errors && validation.errors.length > 0) {
    console.log("Errors:", validation.errors);
  }
}

main().catch(console.error);
